### DISCUSSION TITLE
Textadept 12.5 beta 3

### DISCUSSION THREAD

**Discussion by orbitalquark** [orbitalquark/textadept#572] [General]
Textadept 12.5 beta 3 is released from https://github.com/orbitalquark/textadept/releases/tag/textadept_12.5_beta_3

The biggest change is the renaming of the 'ansi_c', 'dmd', and 'rstats' lexers to 'c', 'd', and 'r', respectively. Custom language-specific keys, snippets, compile/run commands, etc. will be affected. Other than that, macOS support has been greatly improved, and lots of longstanding bugs have been fixed (not all of them though).

Bugfixes:

- Inherit entire login shell environment (including `$PATH`) on macOS without *~/.osx_env.sh*.
- Fixed regression in `reset()`.
- Fixed quick open and find in files filters to not ignore directories like *.github*.
- Fixed slow-pasting of text into the terminal version.
- Fixed rectangular selections not extending beyond end-of-line when virtual space is enabled.
- Fail gracefully if terminal input cannot be read in the terminal version.
- Fixed incremental find spamming find history.
- Fixed insertion of zero-length snippets.
- Fixed inability to quit the terminal version if an initialization error occured.
- Fixed rebinding keys defined in the macOS menubar.
- Support more movement keys in the command entry on macOS.
- Fixed lack of padding in GTK list dialogs.
- Fixed inability to show the first buffer tab in the Qt version when using `ui.SHOW_ALL_TABS`.
- Format: added [`format.prefix_map`][] for formatting paragraphs.
- LSP: fixed autocompletion lists when items have spaces in them.
- LSP: Lua language server recognizes conventional `M` identifier as the current module.
- LSP: lazy-load Textadept API documentation for the Lua command entry.
- LSP: do not scan Textadept's `_HOME` more than once.
- LSP: fixed Lua language server's Textadept `_HOME` detection on macOS.

Changes:

- Updated macOS icon to match Big Sur icon style.
- Updated call tip background to match default background in themes.
- Updated to Qt 6.7 on Windows, which uses Windows 11 theme widgets.
- Updated Brazilian Portuguese translation.
- Renamed 'ansi_c', 'dmd', and 'rstats' lexers to 'c', 'd', and 'r', respectively.
- Language-specific keys have changed (e.g. use `keys.c` and `snippets.c` now).
- Emit `events.VIEW_BEFORE_SWITCH` on `view:split()`.
- Enable `view.caret_line_visible_always`, but only on the current view when Textadept has focus.
- Added `select` option to `ui.dialogs.list()`.
- Use a proper label widget for the command entry label instead of a Scintilla margin.
- Increased terminal version Find & Replace history to 100 items.
- Remove duplicates from Find & Replace history.
- Added Edit > Deselect menu item and key binding.
- Format: use Homebrew's `gfmt` for formatting paragraphs on macOS.
- Updated to [CDK][] 5.0-20240619.

[`format.prefix_map`]: https://github.com/orbitalquark/textadept-format/?tab=readme-ov-file#formatprefix_map-table
[CDK]: https://invisible-island.net/cdk/

**Comment by snoopy** [comment]
The renaming of lexers seems to have broken something in textredux. Typing any of the lexer names (c, d, r) now results in the keychain being activated. Seems like the if clause does not activate in this case because while `_lexer` is `true`, `prefix` is also `true` thus failing the check and activating the keychain path. Textredux does some wonky stuff with key modes so I am not sure yet if this can be fixed. Is there any way to influence the `prefix` parameter from "outside"?

```lua
if not prefix and key._lexer then return PROPAGATE end -- typed key matches lexer name
```

**Reply by rgieseke** [reply]
Huh, good to know. I haven't yet updated to the latest beta.

**Reply by snoopy** [reply]
@rgieseke I think upstream textredux is doubly broken because you are also missing this change https://github.com/snoopy/textredux/commit/ab670ec395d7ed301bf8e80e96ea194424f0d25b

**Reply by orbitalquark** [reply]
I'm not quite sure I understand your question.

`prefix` is either a lexer name (`buffer:get_lexer(true)`), `nil` (if no language-specific key was found), or the current key mode (`keys.MODE`): https://github.com/orbitalquark/textadept/blob/bcd96c8a2528a7d8edac25effbecc9425d813e12/core/keys.lua#L197-L202

Is textredux redefining the language-specific key containers for a given language? (i.e. is it overwriting https://github.com/orbitalquark/textadept/blob/bcd96c8a2528a7d8edac25effbecc9425d813e12/core/keys.lua#L216?)

**Reply by rgieseke** [reply]
I don't think Textredux is doing anything with lexer specific keys.
What I found is that the code for `c`, `d`, `'r` isn't reaching Textredux' `on_char_added` function, other keys work.

https://github.com/rgieseke/textredux/blob/c68ec44314febbf8bc3d761079918ac75c8e026b/core/buffer.lua#L213

**Reply by orbitalquark** [reply]
Okay, this should be fixed by https://github.com/orbitalquark/textadept/commit/13228a03a0e10d7aa11537c84ec0d2d85dc3811c. Sorry about that.

**Reply by rgieseke** [reply]
Wicked, thanks for the quick fix! Works with Textredux now.
